tools/castle-editor/code/framedesign.pas

```diff
@@ -1776,6 +1786,114 @@ procedure TDesignFrame.CastleControlUpdate(Sender: TObject);
+var
+  ShellList: TCastleShellListView;
+  SelectedFileName: String;
+  SelectedURL: String;
+begin
+  if Source is TCastleShellListView then
+  begin
+    ShellList := TCastleShellListView(Source);
+    SelectedFileName := ShellList.GetPathFromItem(ShellList.Selected);
+    SelectedURL := FilenameToURISafe(SelectedFileName);
+
+    Accept := TFileFilterList.Matches(LoadScene_FileFilters, SelectedURL);
+  end else
+    Accept := false;
+end;
+
+procedure TDesignFrame.CastleControlDragDrop(Sender, Source: TObject; X, Y: Integer);
+var
+  ShellList: TCastleShellListView;
+  SelectedFileName: String;
+  SelectedURL: String;
+  Scene: TCastleScene;
+  UI: TCastleUserInterface;
+  Viewport: TCastleViewport;
+  ScenePos: TVector3;
+  RayOrigin, RayDirection: TVector3;
+  RayHit: TRayCollision;
+  Distance: Single;
+  OldPickable: Boolean;
+
+  function Is2DAsset(URL : String): Boolean;
+  var
+    MimeType: String;
+  begin
+    MimeType := URIMimeType(URL);
+    Result := LoadImage_FileFilters.Matches(URL) or
+      (MimeType = 'application/json') or
+      (MimeType = 'application/x-starling-sprite-sheet') or
+      (MimeType = 'application/x-plist') or
+      (MimeType = 'application/x-cocos2d-sprite-sheet');
+  end;
+
+begin
+  if Source is TCastleShellListView then
+  begin
+    ShellList := TCastleShellListView(Source);
+    SelectedFileName := ShellList.GetPathFromItem(ShellList.Selected);
+    SelectedURL := MaybeUseDataProtocol(FilenameToURISafe(SelectedFileName));
+
+    if not TFileFilterList.Matches(LoadScene_FileFilters, SelectedURL) then
+      Exit;
+
+    UI := FDesignerLayer.HoverUserInterface(Vector2(X, Y));
+    if not (UI is TCastleViewport) then
+      Exit;
+
+    Viewport := TCastleViewport(UI);
+    try
+      Scene := AddComponent(Viewport.Items, TCastleScene, nil) as TCastleScene;
+
+      if Is2DAsset(SelectedURL) then
+        Scene.Setup2D;
+
+      Scene.URL := SelectedURL;
+
+      { Make gizmos not pickable when looking for new scene position,
+        because ray can hit on gizmo. }
+      OldPickable := VisualizeTransformSelected.Pickable;
+      try
+        VisualizeTransformSelected.Pickable := false;
+        Viewport.PositionToRay(Vector2(X, CastleControl.Height - Y), true, RayOrigin, RayDirection);
+        RayHit := Viewport.Items.WorldRay(RayOrigin, RayDirection);
+      finally
+        VisualizeTransformSelected.Pickable := OldPickable;
+      end;
+      if RayHit <> nil then
+      begin
+        Distance := RayHit.Distance;
+        FreeAndNil(RayHit);
+      end else
+      begin
+        // if we don't hit any other scene set Distance to default
+        if Viewport.Camera.ProjectionType = ptOrthographic then
+        begin
+          // use camera position for 2D (minimum 100 distance)
+          if Viewport.Camera.Position.Z > 100 then
+            Distance := Viewport.Camera.Position.Z
+          else
+            Distance := 100;
+        end else
+          Distance := 20;
```

Comment:
I fear such hardcoded values, but I see no other choice here :) But make it 10 instead of 20 here -- it seems more "obvious default".

Reply:
I tested 10 and 20. 10 also works OK on our default 3d template. However, I thought the greater value would be safer. So I doubled it up. Of course we can change it to 10 (I'll do that). Maybe that should be in project configuration in the future.